pkg/custom-provider/query_builder.go

```diff
@@ -0,0 +1,205 @@
+package provider
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"strings"
+	"text/template"
+
+	"k8s.io/apimachinery/pkg/selection"
+
+	prom "github.com/directxman12/k8s-prometheus-adapter/pkg/client"
+)
+
+// QueryBuilder provides functions for generating Prometheus queries.
+type QueryBuilder interface {
+	BuildSelector(seriesName string, groupBy string, groupBySlice []string, queryParts []queryPart) (prom.Selector, error)
+}
+
+type queryBuilder struct {
+	metricsQueryTemplate *template.Template
+}
+
+// NewQueryBuilder creates a QueryBuilder.
+func NewQueryBuilder(metricsQuery string) (QueryBuilder, error) {
+	metricsQueryTemplate, err := template.New("metrics-query").Delims("<<", ">>").Parse(metricsQuery)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse metrics query template %q: %v", metricsQuery, err)
+	}
+
+	return &queryBuilder{
+		metricsQueryTemplate: metricsQueryTemplate,
+	}, nil
+}
+
+func (n *queryBuilder) BuildSelector(seriesName string, groupBy string, groupBySlice []string, queryParts []queryPart) (prom.Selector, error) {
+	//Convert our query parts into the types we need for our template.
+	exprs, valuesByName, err := n.processQueryParts(queryParts)
+
+	if err != nil {
+		return "", err
+	}
+
+	args := queryTemplateArgs{
+		Series:            seriesName,
+		LabelMatchers:     strings.Join(exprs, ","),
+		LabelValuesByName: valuesByName,
+		GroupBy:           groupBy,
+		GroupBySlice:      groupBySlice,
+	}
+
+	selector, err := n.createSelectorFromTemplateArgs(args)
+	if err != nil {
+		return "", err
+	}
+
+	return selector, nil
+}
+
+func (n *queryBuilder) createSelectorFromTemplateArgs(args queryTemplateArgs) (prom.Selector, error) {
+	//Turn our template arguments into a Selector.
+	queryBuff := new(bytes.Buffer)
+	if err := n.metricsQueryTemplate.Execute(queryBuff, args); err != nil {
+		return "", err
+	}
+
+	if queryBuff.Len() == 0 {
+		return "", fmt.Errorf("empty query produced by metrics query template")
+	}
+
+	return prom.Selector(queryBuff.String()), nil
+}
+
+func (n *queryBuilder) processQueryParts(queryParts []queryPart) ([]string, map[string][]string, error) {
+	//We've take the approach here that if we can't perfectly map their query into a Prometheus
```

Comment:
comments should have a space between the `//` and the text (e.g. `// this is a comment`, not `//this is a comment`)